nixos/modules/services/x11/clight.nix

```diff
@@ -0,0 +1,111 @@
+{ config, pkgs, lib, ... }:
+
+with lib;
+
+let
+  cfg = config.services.clight;
+
+  toConf = v:
+    if builtins.isFloat v then toString v
+    else if isInt v       then toString v
+    else if isBool v      then boolToString v
+    else if isString v    then ''"${escape [''"''] v}"''
+    else if isList v      then "[ " + concatMapStringsSep ", " toConf v + " ]"
+    else abort "clight.toConf: unexpected type (v = ${v})";
+
+  clightConf = pkgs.writeText "clight.conf"
+    (concatStringsSep "\n" (mapAttrsToList
+      (name: value: "${toString name} = ${toConf value};")
+      (filterAttrs
+        (_: value: value != null)
+        cfg.settings)));
+in {
+  options.services.clight = {
+    enable = mkOption {
+      type = types.bool;
+      default = false;
+      description = ''
+        Whether to enable clight or not.
+      '';
+    };
+
+    temperature = {
+      day = mkOption {
+        type = types.int;
+        default = 5500;
+        description = ''
+          Colour temperature to use during the day, between
+          <literal>1000</literal> and <literal>25000</literal> K.
+        '';
+      };
+      night = mkOption {
+        type = types.int;
+        default = 3700;
+        description = ''
+          Colour temperature to use at night, between
+          <literal>1000</literal> and <literal>25000</literal> K.
+        '';
+      };
+    };
+
+    settings = let
+      validConfigTypes = with types; either int (either str (either bool float));
+    in mkOption {
+      type = with types; attrsOf (nullOr (either validConfigTypes (listOf validConfigTypes)));
+      default = {};
+      description = ''
+        Additional configuration to extend clight.conf.
+      '';
+    };
+  };
+
+  config = mkIf cfg.enable {
+    boot.kernelModules = [ "i2c_dev" ];
+    environment.systemPackages = with pkgs; [ clight clightd ];
+    services.dbus.packages = with pkgs; [ clight clightd ];
+    services.upower.enable = true;
+
+    services.clight.settings = {
+      latitude = mkDefault config.location.latitude;
+      longitude = mkDefault config.location.longitude;
```

Comment:
Shouldn't these options only be set when the location provider is `manual`? If so it should be
```nix
{
  services.clight.settings = mkMerge [
    { gamma_temp = with cfg.temperature; mkDefault [ day night ]; }
    (mkIf (config.location.provider == "manual") {
      latitude = mkDefault config.location.latitude;
      longitude = mkDefault config.location.longitude;
    })
  ];
}
```

Or alternatively with `optionalAttrs`

Reply:
This should only be a problem when `provider == "geoclue2"` and `latitude` and `longitude` are non-null but it's probably better to include it.

Reply:
Oh I just noticed those being `null` by default, but I think it would be better if their type was just `types.float` without any default. The `provider` option should control whether they are used or not (and if they are used they need to be set anyways).

Reply:
Done.